### PR TITLE
Add RViz Script

### LINKED ISSUE #39: Add RViz2 Visualizations for the 3D Messages.
Pull request #38 adds visualizations for a subset of properties. If more features, like the field markings or left/right goal differentiation are used, visualizations should also be added for them.

### PR DESCRIPTION
**Pull request by Flova**
Adds a script which displays the `soccer_vision_3d_msgs` as RViz2 markers.

image
image

Resolves: #39

### CODE REVIEW

File: soccer_vision_3d_msgs/CMakeLists.txt [hunk 1]
```diff
@@ -32,6 +32,10 @@ rosidl_generate_interfaces(${PROJECT_NAME}
 # export dependencies
 ament_export_dependencies(rosidl_default_runtime)
 
+install(PROGRAMS
+  scripts/visualizer.py
+  DESTINATION lib/${PROJECT_NAME} )
```
Comment (jaagut) [inline soccer_vision_3d_msgs/CMakeLists.txt:37]: What does the `lib/` destination refer to? Where will this end up?

File: soccer_vision_3d_msgs/scripts/visualizer.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+import rclpy
+import time
+from rclpy.node import Node
+from rclpy.duration import Duration
+
+from geometry_msgs.msg import Vector3
+from std_msgs.msg import ColorRGBA
+from visualization_msgs.msg import Marker
+from soccer_vision_3d_msgs.msg import BallArray, FieldBoundary, GoalpostArray, RobotArray
```
Comment (jaagut) [inline soccer_vision_3d_msgs/scripts/visualizer.py:11]: `MarkingArray` is still missing. Do you want to implement this in this PR or later on? In the later case, please open an issue.

File: soccer_vision_3d_msgs/scripts/visualizer.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+from soccer_vision_attribute_msgs.msg import Confidence, Robot as RobotAttribute
+
+
+class SoccerVision3DMsgs2Rviz(Node):
```
Comment (jaagut) [inline soccer_vision_3d_msgs/scripts/visualizer.py]: I really don't like this name, but I do not have much better suggestions:

- SoccerVision3DMarkers (with or without 3D)
- SoccerVision3DRvizMarkers (with or without 3D)
- SoccerMarkers
- SoccerRvizMarkers

File: soccer_vision_3d_msgs/scripts/visualizer.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+        # Create publishers
+        self.marker_publisher = self.create_publisher(Marker, "soccer_vision_3d_msgs", 1)
+
+        # Settings
+        self.ball_diameter = 0.13
+        self.ball_lifetime = int(5e9)
+        self.goal_lifetime = int(5e9)
+        self.obstacle_lifetime = int(5e9)
+        self.robot_lifetime = int(5e9)
```
Comment (jaagut) [inline soccer_vision_3d_msgs/scripts/visualizer.py:29]: Do we want to move these settings into a config file?

File: soccer_vision_3d_msgs/scripts/visualizer.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+        # Ball
+        self.marker_ball = Marker()
+        self.marker_ball.ns = "rel_ball"
+        self.marker_ball.id = 0
+        self.marker_ball.type = Marker.SPHERE
+        self.marker_ball.action = Marker.MODIFY
+        self.marker_ball.scale = Vector3(** dict(zip("xyz", [self.ball_diameter] * 3)))
```
Comment (jaagut) [inline soccer_vision_3d_msgs/scripts/visualizer.py]: Why `*3`?

File: soccer_vision_3d_msgs/scripts/visualizer.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    def conf_to_alpha(self, conf: Confidence):
+        if conf.confidence == Confidence.CONFIDENCE_UNKNOWN:
+            return 1.0
+        return conf.confidence
+
+    def balls_cb(self, msg: BallArray):
+        self.marker_ball.header = msg.header
+        for idx, ball in enumerate(msg.balls):
+            self.marker_ball.pose.position = ball.center
+            self.marker_ball.pose.orientation.w = 1.0
```
Comment (jaagut) [inline soccer_vision_3d_msgs/scripts/visualizer.py]: Why is this necessary?

File: soccer_vision_3d_msgs/scripts/visualizer.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        self.marker_ball.color = ColorRGBA(r=1.0, a=1.0)
+        self.marker_ball.lifetime = Duration(nanoseconds=self.ball_lifetime).to_msg()
+
+        # Goalpost
+        self.marker_goalpost = Marker()
+        self.marker_goalpost.ns = "rel_goal"
+        self.marker_goalpost.id = 0
+        self.marker_goalpost.type = Marker.CYLINDER
+        self.marker_goalpost.action = Marker.MODIFY
+        self.marker_goalpost.scale = Vector3()
```
Comment (jaagut) [inline soccer_vision_3d_msgs/scripts/visualizer.py]: Scale gets set later. This line should be unnecessary.

File: soccer_vision_3d_msgs/scripts/visualizer.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+                self.marker_robot.color.g = 0.0
+                self.marker_robot.color.b = 1.0
+            elif robot.attributes.team == RobotAttribute.TEAM_OPPONENT:
+                self.marker_robot.color.r = 1.0
+                self.marker_robot.color.g = 0.0
+                self.marker_robot.color.b = 0.0
+            elif robot.attributes.team == RobotAttribute.TEAM_UNKNOWN:
+                self.marker_robot.color.r = 0.0
+                self.marker_robot.color.g = 1.0
+                self.marker_robot.color.b = 0.0
```
Comment (jaagut) [inline soccer_vision_3d_msgs/scripts/visualizer.py]: Why not use `ColorRGBA(...)`? This shortens the code a lot

File: soccer_vision_3d_msgs/CMakeLists.txt (hunk 1 above)
Comment (Flova) [reply soccer_vision_3d_msgs/CMakeLists.txt:37]: In the install or build directory. Idk. just copied the CMakeLists.txt from another node.

File: soccer_vision_3d_msgs/scripts/visualizer.py (hunk 3 above)
Comment (Flova) [reply soccer_vision_3d_msgs/scripts/visualizer.py]: I like `SoccerVision3DMarkers`.

File: soccer_vision_3d_msgs/scripts/visualizer.py (hunk 4 above)
Comment (Flova) [reply soccer_vision_3d_msgs/scripts/visualizer.py:29]: Maybe, but this is just a viz script. I doubt people really change this.

File: soccer_vision_3d_msgs/scripts/visualizer.py (hunk 5 above)
Comment (Flova) [reply soccer_vision_3d_msgs/scripts/visualizer.py]: This is needless complicated for
`Vector3(x=self.ball_diameter, y=self.ball_diameter, z=self.ball_diameter)`

File: soccer_vision_3d_msgs/scripts/visualizer.py (hunk 6 above)
Comment (Flova) [reply soccer_vision_3d_msgs/scripts/visualizer.py]: The msg is initialized with zero values. But a quaternion needs to have the length != by definition, so it is standard to set the `w=1`. This is still the identity, but it is a defined value instead of an undefined one.

Comment (Flova) [reply soccer_vision_3d_msgs/scripts/visualizer.py]: One could save some computation and move it to the initialization part at the top.

File: soccer_vision_3d_msgs/scripts/visualizer.py (hunk 7 above)
Comment (Flova) [reply soccer_vision_3d_msgs/scripts/visualizer.py]: Right

File: soccer_vision_3d_msgs/scripts/visualizer.py (hunk 8 above)
Comment (Flova) [reply soccer_vision_3d_msgs/scripts/visualizer.py]: Right, this could be a bit more concise.

File: soccer_vision_3d_msgs/scripts/visualizer.py (hunk 2 above)
Comment (Flova) [reply soccer_vision_3d_msgs/scripts/visualizer.py:11]: This PR only adds viz for the subset of features used by the bit-bots. Other features could be added later on. We can open an issue for this.

File: soccer_vision_3d_msgs/scripts/visualizer.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+from visualization_msgs.msg import Marker
+from soccer_vision_3d_msgs.msg import BallArray, FieldBoundary, GoalpostArray, RobotArray
+from soccer_vision_attribute_msgs.msg import Confidence, Robot as RobotAttribute
+
+
+class SoccerVision3DMarkers(Node):
+    """This node provides RViz markers coresponding to the recognized objects."""
+
+    def __init__(self):
+        super().__init__('SVM2DRviz')
```
Comment (jaagut) [inline soccer_vision_3d_msgs/scripts/visualizer.py]: This does not match the class name anymore. Also, nobody will understand this abbreviation, why not spell it out?

**Review by jaagut — COMMENTED** [review]
Thanks again, for your contribution! Great work!

Would you please provide some information on how to start/run this visualization in the package's README and/or the documentation?
I would appreciate, if you could integrate the nice images from your PR into the documentation. (Not into the README, we want to avoid pictures in this git, right @ijnek?)

**Comment by Flova** [comment]
I will do a separate Pull Request for the dokumentation.

**Review by jaagut — APPROVED** [review]

**Comment by ijnek** [comment]
@Flova thank you for this! (and sorry for the late reply - i'm visintg family, but should be more active from early June)

I like the work, the visualizations look nice!

We should take this out and put this work into another work, so that we keep this package a repository containing pure ROS2 interface packages.

Since we have [ros-sports/soccer_vision_2d_layers](https://github.com/ros-sports/soccer_vision_2d_layers) for the soccer_vision_2d_msgs, maybe we could have something similar, like **ros-sports/soccer_vision_3d_rviz_markers** as a new repo. If you agree, I can go ahead and create the repo and you can copy your work across.

**Comment by jaagut** [comment]
@ijnek Yes, I agree, that this makes more sense in another repo and keep this clean.

**Comment by ijnek** [comment]
@Flova I saw the other day that you suggested exactly what I said about separating the repo (on Discord), but I failed to respond to that. Since we all agree, I've created [ros-sports/soccer_vision_3d_rviz_markers](https://github.com/ros-sports/soccer_vision_3d_rviz_markers). Would you be able to move your work across there? Thanks.

**Review by ijnek — CHANGES_REQUESTED** [review]
See comments above

**Comment by ijnek** [comment]
@Flova I believe you don't have time to port this across to the new repo, so I'm starting to work on it. I previously had a [similar package](https://soccer-visualization.readthedocs.io/en/latest/how-to-visualize-soccer-markers.html) (although its outdated), and some notable changes in that package comapred to what you propose here is:

* I used MarkerArray for Array msgs, instead of Marker
* I published different objects on different topics (eg. /visualization/ball, /visualization/goalposts, etc.), rather than different namespaces. I believe this is the only way you can toggle the visualization on and off for each object (ball, goalpost, etc) in RViz.

I also plan on writing some tests. Do you agree with the suggestions?

**Comment by jaagut** [comment]
I believe, this can now be closed as replaced by ros-sports/soccer_vision_3d_rviz_markers/pull/2
